Witches Melt In Water

Wtoll opened the pull request:
Inspired by The Wizard Of Oz, witches now "melt" (get damaged) when exposed to water. This includes water blocks, rain, and splash potions of water. Witches in water blocks, and rain will be damaged over time, and splash potions of water thrown at witches will damage them once, thus following the same rules as that of Blazes and Endermen.

Commit:
Witches melt in water

Witches will be slowly damaged (like Blazes and Endermen) when in water, and will also be damaged by splash potions of water (like Blazes and Enderman).

Commit:
Merge branch 'master' of https://github.com/fabric-community/spooky-time

Commit:
Cleaned up automatic merge

Removed default github formatting for merges

Commit:
Removed duplicate ZombieEntityMixin listing

Shnupbups commented:
I've considered implementing this myself, but I got caught up when I remembered that Witch Huts spawn in Swamps in the Overworld which contain lots of water, which doesn't make much sense if it hurts them.
I love the idea though. Perhaps we could add a variant of the Witch for the Spooky Dimension.that.behaves like this? Give her green skin and call her a Wicked Witch?

Wtoll commented:
I mean witches are already stranded inside of the Witch Huts because their AI wouldn't let them walk off the ledge so in terms of gameplay it doesn't really affect anything. It's not like you ever see witches just kind of wandering around in swamps, they're always stuck inside of the huts by default. I think the only impacts it would have on gameplay are that it would make witch farms easier to build because there's another way to kill them, and it would provide an alternative method of killing witches while caving.

Shnupbups commented:
Hmm, I suppose. Could you add a check that makes them immune to Witch Water?
Perhaps we could say the Witches concocted Witch Water as a way for them to come into contact with water without harm to them.

Zundrel commented:
I definitely do like this idea, but I agree with Witches should be immune to Witch Water.

Commit:
Witch water is a witch safe alternative to water

Witches are no longer affected by Witch Water. Witch Water can now make bubble columns using Tainted Sand and Bleeding Blocks.

Wtoll commented:
Alright, witch water is now a witch safe version of water. This means that witch water no longer hurts witches, but also that witch water can make bubble columns using tainted sand, and bleeding blocks (a new block modeled after the magma block without the hurting part). Feel free for someone more talented to change textures or implement the bleeding block into world generation.

Commit:
Re-added import statements

IntelliJ automagically removed some import statements, so I added them back

Wtoll commented:
Update there's an issue, wait a bit until I get a patch for it

Commit:
Tainted Sand Falls Again

Tainted Sand is once again a falling block

Wtoll commented:
Okay everything ought to be perfect now, I've tested all the interactions and its stable

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/registry/SpookyEvents.java:
```diff
@@ -0,0 +1,23 @@
+
+public class SpookyEvents {
+	public static void init() {
+
```
Do not leave whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/registry/SpookyEvents.java:
```diff
@@ -0,0 +1,23 @@
+			}
+			return ActionResult.PASS;
+		});
+
```
Do not leave whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java:
```diff
@@ -0,0 +1,38 @@
+			info.cancel();
+		}
+	}
+
```
Remove whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java:
```diff
@@ -0,0 +1,159 @@
+	static {
+		DRAG = Properties.DRAG;
+	}
+}
```
This whole class needs `_1` removed from every variable.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/TaintedSandBlock.java:
```diff
@@ -0,0 +1,38 @@
+		return 2;
+	}
+
+	public void onBlockAdded(BlockState state, World world, BlockPos pos, BlockState blockState_2, boolean bool) {
```
Change blockState_2 to not have _2.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/TaintedSandBlock.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Random;
+
+public class TaintedSandBlock extends FallingBlock {
+
```
Remove whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/BleedingBlock.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Random;
+
+public class BleedingBlock extends Block {
+
```
Remove whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/BleedingBlock.java:
```diff
@@ -0,0 +1,56 @@
+				((ServerWorld)world).spawnParticles(ParticleTypes.CURRENT_DOWN, (double)pos2.getX() + 0.5D, (double)pos2.getY() + 0.25D, (double)pos2.getZ() + 0.5D, 8, 0.5D, 0.25D, 0.5D, 0.0D);
+			}
+		}
+
```
Remove whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java:
```diff
@@ -0,0 +1,159 @@
+		} else {
+			entity_1.onBubbleColumnCollision((Boolean)blockState_1.get(DRAG));
+		}
+
```
Remove whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java:
```diff
@@ -0,0 +1,159 @@
+		if (isStillWater(iWorld_1, blockPos_1)) {
+			iWorld_1.setBlockState(blockPos_1, SpookyBlocks.WITCH_WATER_BUBBLE_COLUMN.getDefaultState().with(DRAG, boolean_1), 2);
+		}
+
```
Remove whitespace here.

Shrimpstress commented on src/main/java/com/fabriccommunity/spookytime/block/WitchWaterBubbleColumnBlock.java:
```diff
@@ -0,0 +1,159 @@
+				world_1.playSound(double_1, double_2, double_3, SoundEvents.BLOCK_BUBBLE_COLUMN_UPWARDS_AMBIENT, SoundCategory.BLOCKS, 0.2F + random_1.nextFloat() * 0.2F, 0.9F + random_1.nextFloat() * 0.15F, false);
+			}
+		}
+
```
Remove whitespace here.

Shrimpstress reviewed (changes requested):
Quite a bit of style issues I noticed. You also need to remove the multiple `.DS_Store` files (and preferably add them to the .gitignore). Otherwise, implementation looks fine, most of my notes are simply on style.

Zundrel commented:
I recommend checking out https://github.com/fabric-community/style-guidelines to prevent style issues like this in the future.

Zundrel commented:
Most of your issues overall were with whitespace, so the style guidelines have been updated to reflect our stance on when to not use whitespace.

Commit:
Update the code to follow the style guidelines

Removed extra whitespace, and updated variable names to make the code follow the style guidelines

Commit:
Removed Legacy .DS_Store Files

Wtoll commented:
@Zundrel , everything ought to follow the style guidelines now, and I made sure to force remove those .DS_Store files

Shrimpstress reviewed (approved):
Everything looks great to me now.

Shnupbups committed:
Merge branch 'master' into master

Juuxel commented on src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java:
```diff
@@ -0,0 +1,26 @@
+public class ThrownPotionEntityMixin {
+	@Inject(method = "doesWaterHurt", at = @At("RETURN"))
+	private static boolean doesWaterHurt(LivingEntity livingEntity_1, CallbackInfoReturnable<Boolean> info) {
+		return livingEntity_1 instanceof EndermanEntity || livingEntity_1 instanceof BlazeEntity || livingEntity_1 instanceof WitchEntity;
```
The injection method must be a void method and you should use `info.setReturnValue()` instead of `return`.

Juuxel commented on src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java:
```diff
@@ -0,0 +1,37 @@
+
+@Mixin(WitchEntity.class)
+public abstract class WitchEntityMixin {
+	@Inject(method = "tickMovement", at = @At("HEAD"), cancellable = false)
```
`cancellable = true`

Juuxel commented on src/main/resources/assets/spookytime/textures/block/bleeding_block.png.mcmeta:
```diff
@@ -0,0 +1,11 @@
+      2
+    ]
+  }
+}
```
Tabs or 4 spaces for indentation.

Juuxel commented on src/main/resources/assets/spookytime/lang/en_us.json:
```diff
@@ -47,6 +47,7 @@
 	"block.spookytime.soul_glass_pane": "Soul Glass Pane",
 	"block.spookytime.spookium_ore": "Spookium Ore",
 	"block.spookytime.spookium_block": "Spookium Block",
+  "block.spookytime.bleeding_block": "Bleeding Block",
```
This should be indented with a tab.

Juuxel reviewed (changes requested):
There are some style and mixin issues.

Also, could you add the missing `@Override` annotations to the overridden block methods?

Commit:
Rewrote ThrownPotionEntityMixin Functionality

Rewrote the ThrownPotionEntityMixin functionality so it completely redesigns the way splash potions of water function. It now has a registry system inside of the class so that any mob can be registered to be affected by splash potions of water.

Commit:
Merge branch 'master' of https://github.com/Wtoll/spooky-time

Commit:
Fixed merge conflicts

Commit:
Merged upstream into master

Wtoll commented:
@Juuxel that ought to be everything. I finally read the information about mixins and as a result completely rewrote the ThrownPotionEntityMixin functionality to hopefully be more functional.

Juuxel commented on src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java:
```diff
@@ -1,26 +1,65 @@
+			}
+		}
+		return true;
 	}
```
Do any of these overwrite or override a vanilla method?

Shrimpstress replied:
I don’t think this is even a mixin anymore, I see no injections, redirections, shadowing, modifying variables, etc.

Shrimpstress replied:
In fact I don’t think the methods in this class are even used anywhere else, it sort of seems pointless overall.

Wtoll replied:
Kind of, so damageEntitiesHurtByWater overwrites a method in ThrownPotionEntity, but the ThrownPotionEntityMixin class by Java's standards only extends ThrownEntity so it would throw an error if the @Override annotation was added. In terms of functionality however it does entirely rewrite the way that the original method in ThrownPotionEntity.class functions, but it does it in a safe-ish way because the only time anyone else would want to overwrite that method is if they wanted to add a new entity to be affected by it, but that is handled already by giving the ThrownPotionEntity.class a public method to register new entities via the mixin. Hopefully that made sense.

Shrimpstress replied:
Then use `@Shadow` to get an instance of the list in ThrownPotionEntity and add your entries to it. No need to reinvent the wheel.

Wtoll replied:
ThrownPotionEntity doesn't utilize a list by default, it's all new functionality. Minecraft by default hard codes it as a static method which is why all of those methods had to be added.

Shrimpstress replied:
Then where is this list even being used. You don’t mixin anywhere to use it.

Shrimpstress replied:
And overriding methods is very bad practice, this could easily be done via injection.

Wtoll replied:
The list is a new variable that will get mixed into ThrownPotionEntity, from there the new public method registerHurtByWater allows classes that extend LivingEntity to be registered to be affected by splash potions of water. When a splash potions of water is thrown ThrownEntityPotion already calls damageEntitiesHurtByWater to damage them. That method has been completely rewritten to trigger the new isHurtByWater method to check if the targeted entity is registered in the ThrownPotionEntity's list of entities that are hurt by water.

Shrimpstress replied:
My thing is I don’t see a scenario where we in this mod would want to add functionality to other mobs to be hurt by water potion. Additionally, this breaks usage other mods from changing what mobs are hurt by water, and completely redoes vanilla behavior. This approach I am not very fond of, there was nothing wrong with your previous approach other than returning values in your injection the wrong way, and you could even use a list in that scenario as well.

Juuxel replied:
Please don't overwrite (mixin's `@Overwrite` is assumed if no other annotation is present) target class methods in mixins. That causes crashing incompatibilities when somebody tries to inject to that method. Fabric mods tend to use cancellable injections at `HEAD` that always cancel after the logic, but if you can avoid overwriting, you should.

Wtoll replied:
The issue with using injection is that the original method first asks the world to only send it entities that meet the existing static variable. So if I were to only use injection that call to the world would still be using the old static functionality and not the new dynamic registry functionality.

Juuxel replied:
As I said, you can cancel right after your logic to soft-overwrite the method.

Shrimpstress replied:
Simply put, I will not approve this PR unless we do not Overwrite any methods and only use ways of adding this functionality that do not overwrite vanilla behavior or break functionality with other mods.

Shrimpstress replied:
That is simply unacceptable.

Commit:
Reverted ThrownPotionEntityMixin functionality

Wtoll commented:
@Juuxel Okay that ought to be back to the way it was

valoeghese commented:
I think this would be better for a custom entity. Someone could fork your fork and start working on a custom version of the witch that has this effect.

Wtoll commented:
> I think this would be better for a custom entity. Someone could fork your fork and start working on a custom version of the witch that has this effect.

That was a discussion earlier, but @Shnupbups kind of agreed with me that witches are already stranded in their witch huts and so it wouldn't affect gameplay that much. From there @Zundrel recommended we make witch water a witch safe alternative to water, and that's why that is a feature now. My personal opinion is that keeping it as affecting regular witches won't negatively change gameplay much, and with the new addition of witch water bubble columns it actually gives players a reason to visit the spooky dimension to collect witch water, that way they can make their witch farms.

Draylar commented:
I like the idea; I'll be able to review it later tonight. What if witches could also drink water breathing potions & avoid the damage?

Commit:
Witches with water breathing are not damaged by water

Juuxel commented on src/main/java/com/fabriccommunity/spookytime/block/BreadCrumbsBlock.java:
```diff
@@ -1,48 +1 @@
-	public BlockRenderLayer getRenderLayer() {
-		return BlockRenderLayer.CUTOUT;
-	}
-}
```
What happened here?

Shrimpstress replied:
Oh god something really screwed up here.

Wtoll replied:
Yeah that's really weird, and also not at all correct

Juuxel commented on src/main/java/com/fabriccommunity/spookytime/mixin/ThrownPotionEntityMixin.java:
```diff
@@ -0,0 +1,37 @@
+@Mixin(ThrownPotionEntity.class)
+public class ThrownPotionEntityMixin {
+	@Inject(method = "doesWaterHurt", at = @At("RETURN"), cancellable = true)
+	private static void doesWaterHurt(LivingEntity livingEntity_1, CallbackInfoReturnable<Boolean> info) {
```
You should remove the `_1`, so `LivingEntity entity`

Commit:
Fixed the BreadCrumbsBlock file

Commit:
Fixed the variable names in ThrownPotionEntityMixin

Renamed livingEntity_1 to simply entity in the doesWaterHurt mixin method

Commit:
Merge remote-tracking branch 'upstream/master'

Merged fabric-community/spooky-time:master to fix PR merge issues

Wtoll commented:
@Juuxel the variable name ought to be fixed now and the weirdness with the BreadCrumbsBlock has been resolved

vini2003 commented on src/main/resources/assets/spookytime/lang/en_us.json:
```diff
@@ -54,7 +54,12 @@
 	"block.spookytime.soul_glass_pane": "Soul Glass Pane",
 	"block.spookytime.spookium_ore": "Spookium Ore",
 	"block.spookytime.spookium_block": "Spookium Block",
+<<<<<<< HEAD
```
Seems someone missed a merge conflict.

vini2003 commented on src/main/java/com/fabriccommunity/spookytime/mixin/WitchEntityMixin.java:
```diff
@@ -0,0 +1,26 @@
+	@Inject(method = "tickMovement", at = @At("HEAD"), cancellable = true)
+	public void tick(CallbackInfo info) {
+		ActionResult result = WitchTickCallback.EVENT.invoker().tick( (WitchEntity) (Object) this );
+		if(result == ActionResult.FAIL) {
```
`if (condition)` with a space is the convention.

vini2003 reviewed (changes requested):
You missed a merge conflict and one minor style inconsistency. Aside that, seems good!

Commit:
Minor issues resolved

Changed the if statement in WitchEntityMixin to follow the style guidelines, and removed the merge conflict in the language file

vini2003 committed:
Merge branch 'master' into master
